argoproj-labs/argocd-operator
# changes to support the feature apps in any namespaces

**What type of PR is this?**
/kind enhancement

**What does this PR do / why we need it**:
The PR adds support for the feature apps in any namespace added upstream. To do so, the PR adds new command args `--additional-namespace` to ArgoCD ApplicationController and ArgoCD Server components. Also, the PR adds new roles in the respective namespaces for ArgoCD Server to perform actions in the namespace.

**Have you updated the necessary documentation?**


**Which issue(s) this PR fixes**:
https://issues.redhat.com/browse/GITOPS-2341

**How to test changes / Special notes to the reviewer**:
* Run `make install run`
* Create namespaces `argocd-test` and `argocd-1`
* Deploy an ArgoCD instance using below yaml file
```
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: example-argocd
spec:
  controller:
    sourceNamespaces:
    - argocd
    - argocd-1
  notifications:
    enabled: True
    env:
      - name: foo
        value: bar
  server:
    ingress:
      enabled: true
    sourceNamespaces:
    - argocd-test
    - argocd-1
```
* Check if the ArgoCD pods came up successfully using command `kubectl get pods | grep example-argocd`
* Check whether the new roles were created in namespaces `argocd-test` and `argocd-1`
* Check if the namespaces `argocd-test` and `argocd-1` have a new label `argocd.argoproj.io/managed-by-cluster-argocd`

* e2e test for the PR.
`kubectl kuttl test ./tests/k8s --config ./tests/kuttl-tests.yaml --test 1-024_validate_apps_in_any_namespace`
